frontends/p4/duplicateHierarchicalNameCheck.cpp

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2024 Cisco Systems, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "duplicateHierarchicalNameCheck.h"
+
+#include "lib/log.h"
+
+namespace P4 {
+
+cstring DuplicateHierarchicalNameCheck::getName(const IR::IDeclaration *decl) {
+    return decl->getName();
+}
+
+const IR::Node *DuplicateHierarchicalNameCheck::postorder(IR::Annotation *annotation) {
+    if (annotation->name != IR::Annotation::nameAnnotation) return annotation;
+
+    cstring name = annotation->getName();
+    if (!name.startsWith(".")) {
```

Comment:
Probably it would be better to use `absl::StrJoin` here, something like this:
`absl::StrCat(absl::StrJoin(stack, "."), name.string_view())`. Though this might break up as we'd need to explicitly cast from `cstring` to `string_view`, then we'd need to provide lambda to `StrJoin`.

Reply:
I would like to address this comment, but my C++-fu is pretty weak these days, even with your code snippet and other tips.  If you or someone else has the time to provide a specific working replacement, I am happy to use it.

Reply:
@asl As of commit 25 on this PR, I have tried to address 2 of your 3 comments, and I think it is ready for a fresh review.  If you have more specific suggestions for code changes on this commentt, which I have not addressed as mentioned in my comment above, please let me know.

Reply:
So, does the following work?

```c++
name = absl::StrCat(absl::StrJoin(stack, ".", [](std::string *out, cstring s) {
                                 absl::StrAppend(out, s.string_view());
                             }), name.string_view())
```

Yes, we have not finished the cstring-related implicit conversion saga (I need to find another continuous chunk of time to convert everything...), so we cannot make `cstring` implicitly converted to `string_view` (yet), so these `.string_view()` calls are sadly necessary.

Reply:
addressed in commit 26

Reply:
I will try to keep an eye out for when that PR is merged in, but feel free to ping me when it is, in case I miss it.